server/src/main/java/org/apache/iotdb/db/query/executor/PreviousFillExecutor.java

```diff
@@ -0,0 +1,208 @@
+import org.apache.iotdb.tsfile.read.reader.chunk.ChunkReader;
+
+import java.io.IOException;
+import java.util.*;
+import org.apache.iotdb.tsfile.utils.TsPrimitiveType;
+
+public class PreviousFillExecutor {
+
+  private final Path seriesPath;
+  private final TSDataType dataType;
+  private final QueryContext context;
+  private long queryTime;
+
+  private final Filter timeFilter;
+
+  /*
+   * file cache
+   */
+  private final List<TsFileResource> seqFileResource;
+  private final PriorityQueue<TsFileResource> unseqFileResource;
+
+  /*
+   * chunk cache
+   */
+  private ChunkMetaData lastSeqChunkMetaData;
+  private final List<ChunkMetaData> chunkMetadatas;
+
+  public PreviousFillExecutor(Path seriesPath, TSDataType dataType, QueryContext context,
+      QueryDataSource dataSource, Filter timeFilter, Filter valueFilter, TsFileFilter fileFilter,
+      long queryTime) {
+    this.seriesPath = seriesPath;
+    this.dataType = dataType;
+    this.context = context;
+    QueryUtils.filterQueryDataSource(dataSource, fileFilter);
+    this.seqFileResource = dataSource.getSeqResources();
+    this.unseqFileResource = sortUnSeqFileResourcesInDecendingOrder(dataSource.getUnseqResources());
+    this.chunkMetadatas = new ArrayList<>();
+    this.timeFilter = timeFilter;
+    this.queryTime = queryTime;
+  }
+
+  public TimeValuePair getLastPoint() throws IOException {
+    UnpackAllOverlappedFilesToChunkMetadatas();
+
+    TimeValuePair lastPoint = new TimeValuePair(Long.MIN_VALUE, null);
+    long lastVersion = 0;
+    while (!chunkMetadatas.isEmpty()) {
+      ChunkMetaData chunkMetaData = chunkMetadatas.remove(0);
+      TimeValuePair lastChunkPoint = getChunkLastPoint(chunkMetaData);
+      if (shouldUpdate(
+          lastPoint.getTimestamp(), chunkMetaData.getVersion(),
+          lastChunkPoint.getTimestamp(), lastVersion)) {
+        lastPoint = lastChunkPoint;
+      }
+    }
+    return lastPoint;
+  }
+
+  private TimeValuePair getChunkLastPoint(ChunkMetaData chunkMetaData) throws IOException {
+    Statistics chunkStatistics = chunkMetaData.getStatistics();
+    if (!timeFilter.satisfy(chunkStatistics)) {
+      return null;
+    }
+    if (containedByTimeFilter(chunkStatistics)) {
+      return constructLastPair(
+          chunkStatistics.getEndTime(), chunkStatistics.getLastValue(), dataType);
+    } else {
+      List<IPageReader> pageReaders = unpackChunkReaderToPageReaderList(chunkMetaData);
+      TimeValuePair lastPoint = new TimeValuePair(0, null);
+      for (IPageReader pageReader : pageReaders) {
+        TimeValuePair lastPagePoint = getPageLastPoint(pageReader);
+        if (lastPoint.getTimestamp() < lastPagePoint.getTimestamp()) {
+          lastPoint = lastPagePoint;
+        }
+      }
+      return lastPoint;
+    }
+  }
+
+  private TimeValuePair getPageLastPoint(IPageReader pageReader) throws IOException {
+    Statistics pageStatistics = pageReader.getStatistics();
+    if (!timeFilter.satisfy(pageStatistics)) {
```

Comment:
the pageReader must satisfy this timeFilter, otherwise it will be removed when unpack the ChunkMetadata. This check is not needed.